File: node-client/src/pinus/src/Network.ts
import { EventEmitter } from 'events'
import { Structs } from 'struct-routes';
import { INetworkHandler } from './INetworkHandler';
import { Message, MessageType, Package, Protocol, Client, ERR_CONNECT_TIMEOUT, EVENT_BEENKICKED, EVENT_CLOSED, EVENT_CONNECTED, EVENT_ERROR, EVENT_HANDSHAKEERROR, EVENT_HANDSHAKEOVER, EVENT_RECONNECTED, timestr } from './internal'

const NETWORK_LOG = true;

const JS_WS_CLIENT_TYPE = 'js-websocket';
const JS_WS_CLIENT_VERSION = '0.0.1';
const RES_OK = 200;
const RES_FAIL = 500;
const RES_OLD_CLIENT = 501;
const MAX_RECONNECT_ATTEMPTS = 5;

export interface IHandshakeBuffer {
    sys: {
        type: string,
        version: string,
        rsa: any
    };
    user: any;
}

export class Network extends EventEmitter implements INetworkHandler {

    // --- instances begin ---
    protected static _instances: Map<string, Network> = new Map<string, Network>();
    public static clear(key: string): void {
        const old = Network._instances.get(key);
        if (old) old.disconnect();
        Network._instances.delete(key);
    }

    public static get(key: string): Network {
        let instance = Network._instances.get(key);
        if (instance) return instance;

        instance = new Network(key);
        Network._instances.set(key, instance);
        return instance;
    }
    // --- instances end ---

    // --- EventTarget begin ---
    protected event(name: string, data: any = null) {
        if (this.listenerCount(name)) {
            this.emit(name, data);
        }
        else {
            NETWORK_LOG && console.warn(timestr(), 'pinus event [' + name + '] no listener')
        }
    }
    // --- EventTarget end ---

    // --- ISchedulable begin ---
    public id?: string;
    public uuid?: string;
    // --- ISchedulable end ---

    protected _client: Client = null;
    public get client(): Client { return this._client; }

    public get Url(): string { return this._client?.Url; };
    public set Url(v: string) { if (this._client) this._client.Url = v; }

    public AutoReconnect: boolean = true;

    protected _reconnecting: boolean = false;
    protected _reconnectAttempts: number = 0;

    protected _heartbeatPassed: number = 0;
    protected _heartbeatInterval: number = 0;
    protected _heartbeatTimeout: number;
    protected _shouldHeartbeat: boolean = false;

    protected _requestId: number = 1;
    public get uniqueRequestId(): number {
        this._requestId++;
        if (this._requestId >= 40000) this._requestId = 1;
        return this._requestId;
    }

    // Map from request id to route
    protected _requestRouteMap: Map<number, number | string> = new Map();
    // callback from request id
    protected _requestCallbackMap: Map<number, (data: any) => void> = new Map();

    protected _handshakeBuffer: IHandshakeBuffer = {
        sys: {
            type: JS_WS_CLIENT_TYPE,
            version: JS_WS_CLIENT_VERSION,
            rsa: {}
        },
        user: {}
    };

    protected _routeMap: Map<string, number> = null;
    protected _routeMapBack: Map<number, string> = null;

    protected constructor(key: string = null) {
        super();
        this.id = key;
        this.uuid = key;
        this._client = new Client(this);
        setTimeout(() => {
            this.heartbeatCheck(0.1);
        }, 10);
    }

    // --- Socket begin ---
    public onOpen() {
        if (this._reconnecting) {
            this.event(EVENT_RECONNECTED);
        }
        this.event(EVENT_CONNECTED);
        this.resetReconnect();
        var obj = Package.encode(Package.TYPE_HANDSHAKE, Protocol.strencode(JSON.stringify(this._handshakeBuffer)));
        this.client.sendBuffer(obj);
    }

    public onRecv(data: any) {
        this.processPackage(Package.decode(data));

        // new package arrived, update the heartbeat timeout
        this.renewHeartbeatTimeout();
    }

    public onErr(err: any) {
        this.event(EVENT_ERROR, err);
    }

    public onClose() {
        this.event(EVENT_CLOSED);
        if (this.AutoReconnect && this._reconnectAttempts < MAX_RECONNECT_ATTEMPTS) {
            this._reconnecting = true;
            this._reconnectAttempts++;
            this._connect();
        }
    }

    public connectTimeout() {
        this.event(EVENT_ERROR, ERR_CONNECT_TIMEOUT);
        if (this._reconnectAttempts < MAX_RECONNECT_ATTEMPTS) {
            this._reconnecting = true;
            this._reconnectAttempts++;
            this._connect();
        }
    }
    // --- Socket end ---

    protected resetReconnect() {
        this._reconnecting = false;
        this._reconnectAttempts = 0;
    }

    protected renewHeartbeatTimeout() {
        this._heartbeatPassed = 0;
    }

    protected handshake(data: ArrayBuffer) {
        const d: { code: number, sys: { heartbeat: number, dict: any } } = JSON.parse(Protocol.strdecode(data));
        if (d && d.code === RES_OLD_CLIENT) {
            this.event(EVENT_HANDSHAKEERROR);
            return;
        }

        if (d && d.code !== RES_OK) {
            this.event(EVENT_HANDSHAKEERROR);
            return;
        }

        if (d && d.sys && d.sys.heartbeat) {
            this._heartbeatInterval = d.sys.heartbeat;              // heartbeat interval
            this._heartbeatTimeout = this._heartbeatInterval * 2;   // max heartbeat timeout
        } else {
            this._heartbeatInterval = 0;
            this._heartbeatTimeout = 0;
        }

        if (d && d.sys) {
            const dict = d.sys.dict;

            // Init compress dict
            if (dict) {
                this._routeMap = new Map();
                this._routeMapBack = new Map();
                for (const key in dict) {
                    if (Object.prototype.hasOwnProperty.call(dict, key)) {
                        const value: number = dict[key];
                        this._routeMap.set(key, value);
                        this._routeMapBack.set(value, key);
                    }
                }
            }
        }

        this.client.sendBuffer(Package.encode(Package.TYPE_HANDSHAKE_ACK));
        this.event(EVENT_HANDSHAKEOVER);
    }

    protected heartbeat(data: ArrayBuffer) {
        if (!this._heartbeatInterval) {
            // no heartbeat
            return;
        }

        this._shouldHeartbeat = true;
    }

    protected heartbeatCheck(dt: number) {
        if (!this._heartbeatInterval) return;

        if (!this._client.isConnected) {
            this._heartbeatPassed = 0;
            return;
        }

        this._heartbeatPassed += dt;

        if (this._shouldHeartbeat) {
            if (this._heartbeatPassed > this._heartbeatInterval) {
                this.client.sendBuffer(Package.encode(Package.TYPE_HEARTBEAT));
                this.renewHeartbeatTimeout();
            }
            return;
        }

        if (this._heartbeatPassed > this._heartbeatTimeout) {
            console.error('pinus server heartbeat timeout');
            if (this._reconnectAttempts < MAX_RECONNECT_ATTEMPTS) {
                this._reconnecting = true;
                this._reconnectAttempts++;
                this._connect();
            }
        }
    }

    protected onData(data: ArrayBuffer) {
        const msg = this.decode(data);

        if (!msg) {
            console.error('pinus onData decode failed');
            return;
        }
        else {
            NETWORK_LOG && console.log(timestr(), 'recv', msg);
        }

        if (msg.id) {
            // if have a id then find the callback function with the request
            var cb = this._requestCallbackMap.get(msg.id);
            this._requestCallbackMap.delete(msg.id);
            cb && cb(msg.body);
            return;
        }

        // server push message
        this.event(msg.route as string, msg.body);
    }

    onKick(data: ArrayBuffer) {
        data = JSON.parse(Protocol.strdecode(data));
        this.event(EVENT_BEENKICKED, data);
    }

    protected _messageHandlers: Map<number, (data: ArrayBuffer) => void> = null;
    protected get messageHandlers(): Map<number, (data: ArrayBuffer) => void> {
        if (!this._messageHandlers) {
            this._messageHandlers = new Map();
            this._messageHandlers.set(Package.TYPE_HANDSHAKE, this.handshake.bind(this));
            this._messageHandlers.set(Package.TYPE_HEARTBEAT, this.heartbeat.bind(this));
            this._messageHandlers.set(Package.TYPE_DATA, this.onData.bind(this));
            this._messageHandlers.set(Package.TYPE_KICK, this.onKick.bind(this));
        }
        return this._messageHandlers;
    }

    protected processPackage(msgs: { type: number, body?: Uint8Array } | { type: number, body?: Uint8Array }[]) {
        if (Array.isArray(msgs)) {
            for (var i = 0; i < msgs.length; i++) {
                var msg = msgs[i];
                this.messageHandlers.get(msg.type)(msg.body);
            }
        } else {
            this.messageHandlers.get(msgs.type)(msgs.body);
        }
    }

    public connect(url: string, autoReconnect: boolean = true): void {
        this.Url = url;
        this.AutoReconnect = autoReconnect;
        this._connect();
    }

    protected _connect(): void {
        this._disconnect();

        const once = setTimeout(() => {
            this.client.connect();
            clearTimeout(once);
        }, 10);
    }

    public disconnect(): void {
        this.AutoReconnect = false;
        this._disconnect();
    }

    protected _disconnect(): void {
        this._reconnecting = false;
        this.client.close();
    }

    public packProto(data: any, protoStruct: any): Uint8Array {
        let buffer: Uint8Array = null;
        if (data && protoStruct) {
            let message: any = null;
            if (protoStruct) {
                message = protoStruct.create(data);
                buffer = protoStruct.encode(message).finish();
            }
        }
        return buffer;
    }

    public parseProto(buffer: Uint8Array, protoStruct: any): any {
        let decoded: any = null;
        if (buffer && buffer.length && protoStruct) {
            try {
                decoded = protoStruct.decode(buffer);
            }
            catch (e) {
                console.error('pinus proto decode error', e);
                decoded = null;
            }
        }
        return decoded;
    }

    protected decode(data: string | ArrayBuffer): {
        id: number,
        route: number | string,
        body: any,
    } {
        // probuff decode
        const msg = Message.decode(data);

        if (msg.id > 0) {
            msg.route = this._requestRouteMap.get(msg.id);
            this._requestRouteMap.delete(msg.id);
            if (!msg.route) {
                return null;
            }
        }

        let route = msg.route;

        // Decompose route from dict
        if (msg.compressRoute) {
            route = this._routeMapBack.get(route as number);
            if (!route) return null;

            msg.route = route;
        }

        const cmd = Structs.getCmd(route as string);
        msg.body = this.parseProto(msg.body, cmd.server);
        return msg;
    }

    protected encode(reqId: number, route: number | string, msg: any): Uint8Array {
        var type = reqId ? MessageType.REQUEST : MessageType.NOTIFY;

        const cmd = Structs.getCmd(route as string);
        msg = this.packProto(msg, cmd.server);

        var compressRoute = 0;
        if (this._routeMap.has(route as string)) {
            route = this._routeMap.get(route as string);
            compressRoute = 1;
        }

        return Message.encode(reqId, type, compressRoute, route, msg);
    }

    protected _sendMessage(reqId: number, route: number | string, msg: any): void {
        NETWORK_LOG && console.log(timestr(), 'sendMessage', reqId, route, msg);
        const message = this.encode(reqId, route, msg);
        this.client.sendBuffer(Package.encode(Package.TYPE_DATA, message));
    }

    public request(route: string, msg: any, cb: (data: any) => void) {

        route = route || msg.route;
        if (!route) {
            return;
        }

        const reqId = this.uniqueRequestId;

        this._sendMessage(reqId, route, msg);

        this._requestCallbackMap.set(reqId, cb);
        this._requestRouteMap.set(reqId, route);
    }

    public notify(route: number | string, msg: any): void {
        this._sendMessage(0, route, msg);
    }
}

export const network = Network.get('pinus.network.default');
export const on = network.on.bind(network);
export const off = network.off.bind(network);
export const connect = network.connect.bind(network);
export const disconnect = network.disconnect.bind(network);
export const request = network.request.bind(network);
export const notify = network.notify.bind(network);

File: unity-client/Assets/Plugins/Pinus-Unity/Network.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using UnityEngine;

namespace PinusUnity
{
    public class Network : INetworkHandler
    {
        const int RES_OK = 200;
        const int RES_FAIL = 500;
        const int RES_OLD_CLIENT = 501;

        // --- instances begin ---

        protected static Dictionary<string, Network> _instances = new Dictionary<string, Network>();

        public static void Clear(string key)
        {
            Network old = null;
            if (_instances.TryGetValue(key, out old))
            {
                old.Disconnect();
                _instances.Remove(key);
            }
        }

        public static Network Get(string key)
        {
            Network instance = null;
            if (_instances.TryGetValue(key, out instance))
            {
                return instance;
            }

            instance = new Network(key);
            _instances[key] = instance;
            return instance;
        }

        public static Network Default { get { return Get("Default"); } }
        // --- instances end ---


        protected Client Client { get; set; }
        public string Url
        {
            get { return Client.Url; }
            set { if (Client != null) Client.Url = value; }
        }

        protected bool m_AutoReconnect = true;
        public bool AutoReconnect { get => m_AutoReconnect; set => m_AutoReconnect = value; }

        protected bool Reconnecting { get; set; }
        protected int ReconnectAttempt { get; set; }

        protected int m_MaxReconnectAttempts = 3;
        public int MaxReconnectAttempts { get => m_MaxReconnectAttempts; set => m_MaxReconnectAttempts = value; }

        protected float HeartbeatPassed { get; set; }
        protected float HeartbeatInterval { get; set; }
        protected float HeartbeatTimeout { get; set; }
        protected bool ShouldHeartbeat { get; set; }

        protected int m_RequestId = 1;
        public int GenerateUniqueRequestId()
        {
            m_RequestId++;
            if (m_RequestId >= 40000) m_RequestId = 1;
            return m_RequestId;
        }

        // Map from request id to route
        protected Dictionary<int, string> m_RequestRouteMap = new Dictionary<int, string>();

        // callback from request id
        protected Dictionary<int, Action<Google.Protobuf.IMessage>> m_RequestCallbackMap = new Dictionary<int, Action<Google.Protobuf.IMessage>>();

        protected Dictionary<string, int> m_RouteMap = null;
        protected Dictionary<int, string> m_RouteMapBack = null;

        protected Network(string key)
        {
            Client = new Client(this);
            EventBus.Instance.OnFrameUpdated -= HeartbeatCheck;
            EventBus.Instance.OnFrameUpdated += HeartbeatCheck;
        }

        // --- Socket begin ---
        private byte[] m_HandshakeBuff = null;
        private byte[] HandshakeBuff
        {
            get
            {
                if (m_HandshakeBuff == null)
                {
                    HandshakeSendPackage handshake = new HandshakeSendPackage
                    {
                        sys = new HandshakeSendPackageSys
                        {
                            type = "unity",
                            version = "1.0.0",
                            rsa = null
                        },
                        user = null
                    };

                    var str = JsonConvert.SerializeObject(handshake);
                    var obj = Protocol.StrEncode(str);
                    m_HandshakeBuff = Package.Encode(PackageType.Handshake, obj, obj.Length);
                }
                return m_HandshakeBuff;
            }
        }

        public void OnOpen()
        {
            if (Reconnecting)
            {
                EventBus.Instance.Reconnected(Url);
            }

            EventBus.Instance.Connected(Url);

            ResetReconnect();

            Client.SendBuffer(HandshakeBuff);
        }

        public void OnRecv(byte[] data)
        {
            ProcessPackage(data);

            // new package arrived, update the heartbeat timeout
            RenewHeartbeatTimeout();
        }

        public void OnError(Exception err)
        {
            EventBus.Instance.Error(Url, err);
        }

        public void OnClose()
        {
            EventBus.Instance.Closed(Url);

            if (AutoReconnect && ReconnectAttempt < MaxReconnectAttempts)
            {
                Reconnecting = true;
                ReconnectAttempt++;
                Connect();
            }
        }

        public void ConnectTimeout()
        {
            EventBus.Instance.Error(Url, new Exception("Connect Timeout"));
            if (ReconnectAttempt < MaxReconnectAttempts)
            {
                Reconnecting = true;
                ReconnectAttempt++;
                Connect();
            }
        }

        // --- Socket end ---

        protected void ResetReconnect()
        {
            Reconnecting = false;
            ReconnectAttempt = 0;
        }

        protected void RenewHeartbeatTimeout()
        {
            HeartbeatPassed = 0;
        }

        void OnHandshake(byte[] data, int offset, int length)
        {
            var str = Protocol.StrDecode(data, offset, length);
            var d = JsonConvert.DeserializeObject<HandshakeRecvPackage>(str);
            if (d.code == RES_OLD_CLIENT)
            {
                EventBus.Instance.HandshakeError(Url, new Exception("client version not fullfill"));
                return;
            }

            if (d.code != RES_OK)
            {
                EventBus.Instance.HandshakeError(Url, new Exception("handshake fail"));
                return;
            }

            if (d.sys.heartbeat != 0)
            {
                HeartbeatInterval = d.sys.heartbeat;        // heartbeat interval
                HeartbeatTimeout = HeartbeatInterval * 2;   // max heartbeat timeout
            }
            else
            {
                HeartbeatInterval = 0;
                HeartbeatTimeout = 0;
            }

            if (d.sys.dict != null)
            {
                var dict = d.sys.dict;
                m_RouteMap = new Dictionary<string, int>();
                m_RouteMapBack = new Dictionary<int, string>();
                foreach (var pair in dict)
                {
                    m_RouteMap[pair.Key] = pair.Value;
                    m_RouteMapBack[pair.Value] = pair.Key;
                }
            }

            Client.SendBuffer(Package.Encode(PackageType.HandshakeAck));
            EventBus.Instance.HandshakeOver(Url);
        }

        void Heartbeat()
        {
            if (HeartbeatInterval == 0)
            {
                // no heartbeat
                return;
            }

            ShouldHeartbeat = true;
        }

        void HeartbeatCheck()
        {
            if (HeartbeatInterval == 0) return;

            if (!Client.IsConnected)
            {
                HeartbeatPassed = 0;
                return;
            }

            var dt = Time.unscaledDeltaTime;

            HeartbeatPassed += dt;

            if (ShouldHeartbeat)
            {
                if (HeartbeatPassed > HeartbeatInterval)
                {
                    Client.SendBuffer(Package.Encode(PackageType.Heartbeat));
                    RenewHeartbeatTimeout();
                }
                return;
            }

            if (HeartbeatPassed > HeartbeatTimeout)
            {
                Utils.L("Pinus Server Heartbeat Timeout");
                if (ReconnectAttempt < MaxReconnectAttempts)
                {
                    Reconnecting = true;
                    ReconnectAttempt++;
                    Connect();
                }
            }
        }

        void OnData(byte[] bytes, int offset, int length)
        {
            int id = 0;
            MessageType type;
            int routeCode = 0;
            string routeStr = null;
            int bodyOffset = 0;
            int bodyLength = 0;
            Message.Decode(bytes, offset, length, out id, out type, out routeCode, out routeStr, out bodyOffset, out bodyLength);

            if (id > 0)
            {
                if (m_RequestRouteMap.ContainsKey(id))
                {
                    routeStr = m_RequestRouteMap[id];
                    m_RequestRouteMap.Remove(id);
                }
                else
                {
                    Log.E("Pinus Server Response Error: Unknown Request Id: " + id);
                    return;
                }
            }

            // Decompose route from dict
            if (routeStr == null)
            {
                m_RouteMapBack.TryGetValue(routeCode, out routeStr);

                if (routeStr == null)
                {
                    Log.E("Pinus Server Response Error: Unknown Route Code: " + routeCode);
                    return;
                }
            }

            var msg = ProtoCoder.ParseProto(routeStr, bytes, bodyOffset, bodyLength);

            if (msg == null)
            {
                Log.E("Pinus Network OnData Decode Failed");
                return;
            }
            else
            {
                Utils.L("Pinus Network OnData", msg);
            }

            if (id != 0)
            {
                // if have a id then find the callback function with the request
                Action<Google.Protobuf.IMessage> cb = null;
                if (m_RequestCallbackMap.TryGetValue(id, out cb))
                {
                    m_RequestCallbackMap.Remove(id);
                    cb(msg);
                }
                return;
            }

            EventDispatcher.Dispatch(routeStr, msg);
        }

        void OnKick(byte[] data, int offset, int length)
        {
            var d = Protocol.StrDecode(data, offset, length);
            Log.D("Pinus Network OnKick", d);

            EventBus.Instance.BeenKicked(Url);
            Client.Close();
        }

        internal void ProcessPackage(byte[] bytes)
        {
            int offset = 0;
            int length = 0;

            var type = Package.Decode(bytes, ref offset, out length);
            while (type != PackageType.Unknown)
            {
                ProcessMessage(type, bytes, offset, length);
                offset += length;
                type = Package.Decode(bytes, ref offset, out length);
            }
        }

        internal void ProcessMessage(PackageType type, byte[] bytes, int offset, int length)
        {
            switch (type)
            {
                case PackageType.Handshake:
                    OnHandshake(bytes, offset, length);
                    break;
                case PackageType.Heartbeat:
                    Heartbeat();
                    break;
                case PackageType.Data:
                    OnData(bytes, offset, length);
                    break;
                case PackageType.Kick:
                    OnKick(bytes, offset, length);
                    break;
                default:
                    break;
            }
        }

        public void Connect(string url, bool autoReconnect = true)
        {
            Url = url;
            AutoReconnect = autoReconnect;
            Connect();
        }

        protected void Connect()
        {
            DoDisconnect();

            Client.Connect();
        }

        public void Disconnect()
        {
            AutoReconnect = false;
            DoDisconnect();
        }

        protected void DoDisconnect()
        {
            Reconnecting = false;
            Client.Close();
        }

        private byte[] m_DataBuffer = new byte[2 * 1024];
        private byte[] m_SendBuffer = new byte[4 * 1024];
        public void SendMessage<T>(int requestId, string route, T msg) where T : Google.Protobuf.IMessage
        {
            if (String.IsNullOrEmpty(route))
            {
                Log.E("Pinus Network Notify Error: Empty Route");
                return;
            }

            Utils.L("Pinus Network SendMessage", requestId, route, msg);

            var type = requestId > 0 ? MessageType.REQUEST : MessageType.NOTIFY;

            var routeCode = 0;
            if (!m_RouteMap.TryGetValue(route, out routeCode)) routeCode = 0;

            var dataLen = ProtoCoder.PackProto(route, msg, ref m_DataBuffer);
            var offset = 0;

            if (routeCode == 0)
            {
                Message.Encode(requestId, type, route, m_DataBuffer, dataLen, ref m_SendBuffer, ref offset);
            }
            else
            {
                Message.Encode(requestId, type, routeCode, m_DataBuffer, dataLen, ref m_SendBuffer, ref offset);
            }

            Client.SendBuffer(Package.Encode(PackageType.Data, m_SendBuffer, offset));
        }

        public void Request<T>(string route, T msg, Action<T> cb = null) where T : Google.Protobuf.IMessage
        {
            if (String.IsNullOrEmpty(route))
            {
                Log.E("Pinus Network Request Error: Empty Route");
                return;
            }

            var requestId = GenerateUniqueRequestId();
            SendMessage(requestId, route, msg);
            if (cb != null) m_RequestCallbackMap.Add(requestId, (e) => cb((T)e));
            m_RequestRouteMap.Add(requestId, route);
        }

        public void Notify<T>(string route, T data) where T : Google.Protobuf.IMessage
        {
            if (String.IsNullOrEmpty(route))
            {
                Log.E("Pinus Network Notify Error: Empty Route");
                return;
            }

            SendMessage(0, route, data);
        }
    }
}